### PR TITLE
SOLR-16588: set default knn algorithm

### PR DESCRIPTION
**Pull request by alessandrobenedetti**
https://issues.apache.org/jira/browse/SOLR-16588

# Description

Bugfix for Hnsw algorithm

# Solution

Better param validation

# Tests

minor test reformat
# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/core/SchemaCodecFactory.java [hunk 1]
```diff
@@ -124,18 +124,18 @@ public KnnVectorsFormat getKnnVectorsFormatForField(String field) {
             if (fieldType instanceof DenseVectorField) {
               DenseVectorField vectorType = (DenseVectorField) fieldType;
               String knnAlgorithm = vectorType.getKnnAlgorithm();
-              if (knnAlgorithm != null) {
-                if (knnAlgorithm.equals(DenseVectorField.HNSW_ALGORITHM)) {
-                  int maxConn = vectorType.getHnswMaxConn();
-                  int beamWidth = vectorType.getHnswBeamWidth();
-                  return new Lucene94HnswVectorsFormat(maxConn, beamWidth);
-                }
+              if (knnAlgorithm.equals(DenseVectorField.HNSW_ALGORITHM)) {
```
Comment (risdenk) [inline solr/core/src/java/org/apache/solr/core/SchemaCodecFactory.java:127]: nit the following will avoid any NPE if `knnAlgorithm` is ever null.: 
```suggestion
              if (DenseVectorField.HNSW_ALGORITHM.equals(knnAlgorithm)) {
```

File: solr/core/src/java/org/apache/solr/schema/DenseVectorField.java [hunk 2]
```diff
@@ -104,7 +106,7 @@ public void init(IndexSchema schema, Map<String, String> args) {
             .orElse(DEFAULT_SIMILARITY);
     args.remove(KNN_SIMILARITY_FUNCTION);
 
-    this.knnAlgorithm = args.get(KNN_ALGORITHM);
+    this.knnAlgorithm = Optional.ofNullable(args.get(KNN_ALGORITHM)).orElse(DEFAULT_KNN_ALGORITHM);
```
Comment (risdenk) [inline solr/core/src/java/org/apache/solr/schema/DenseVectorField.java:109]: use getOrDefault? https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#getOrDefault-java.lang.Object-V-

```suggestion
    this.knnAlgorithm = args.getOrDefault(KNN_ALGORITHM, DEFAULT_KNN_ALGORITHM);
```

Comment (risdenk) [reply solr/core/src/java/org/apache/solr/schema/DenseVectorField.java:109]: This should cleanup the imports too.

**Review by risdenk — CHANGES_REQUESTED** [review]

**Review by risdenk — APPROVED** [review]

**Comment by alessandrobenedetti** [comment]
thanks @risdenk ! All committed, adding the Changes.txt as soon as I know this is coming in 9.1.1!
Cheers

File: solr/core/src/java/org/apache/solr/core/SchemaCodecFactory.java [hunk 3]
```diff
@@ -124,18 +124,18 @@ public KnnVectorsFormat getKnnVectorsFormatForField(String field) {
...
-              if (knnAlgorithm != null) {
-                if (knnAlgorithm.equals(DenseVectorField.HNSW_ALGORITHM)) {
-                  int maxConn = vectorType.getHnswMaxConn();
-                  int beamWidth = vectorType.getHnswBeamWidth();
-                  return new Lucene94HnswVectorsFormat(maxConn, beamWidth);
-                }
+              if (DenseVectorField.HNSW_ALGORITHM.equals(knnAlgorithm)) {
+                int maxConn = vectorType.getHnswMaxConn();
+                int beamWidth = vectorType.getHnswBeamWidth();
+                return new Lucene94HnswVectorsFormat(maxConn, beamWidth);
```
Comment (henrik242) [inline solr/core/src/java/org/apache/solr/core/SchemaCodecFactory.java:130]: With regards to including this in Solr: 9.1.1:  Doesn't `Lucene94HnswVectorsFormat` require Lucene 9.4?  9.1.1 will stay on Lucene 9.3 afaik.

Comment (alessandrobenedetti) [reply solr/core/src/java/org/apache/solr/core/SchemaCodecFactory.java:130]: You are absolutely right!
Can't we then cherry-pick and change it to Lucene93HnswVectorsFormat for the 9.1.1 branch?
I think it's a nice to have bugfix!

Comment (henrik242) [reply solr/core/src/java/org/apache/solr/core/SchemaCodecFactory.java:130]: I would assume so! :)

**Pull request merged**